### DISCUSSION TITLE
Characterset for the connection

### DISCUSSION THREAD

**Discussion by marcin-bury** [asfernandes/node-firebird-drivers#97] [Q&A] [answered]
Hello

Is there a way to provide the characterset for the connection?
I have legacy database with WIN1250 charset, and I notices that characterset is hardcoded (?) to 'utf8'

TIA
Marcin

**Comment by asfernandes** [comment, marked as answer]
Node uses UTF-8 strings. You can use it even with columns using others charsets.

Firebird will do the conversions.

I don't see a need to let the connection charset be changed.

**Reply by marcin-bury** [reply]
I have asked, because Postman/console.log in VSCode shows:
"Us�uga" instead of "Usługa" for record inserted with Delphi application (with WIN1250 specified for the connection).
Newly inserted values look ok when browsed via web application but in Delphi application the text is glitched.
So probably due to different charset used for both connections.
I would change charset of the database to utf8 if it would be possible, but it is not at the moment.

Marcin

**Reply by asfernandes** [reply]
Maybe your records are incorrectly encoded?

Say, you inserted UTF-8 data without UTF8 connection charset into the WIN1250 fields.

If then, you should correct them.

Firebird should convert things correctly when you use the correct connection charset - which exists for that.

Or your console is configured incorrectly.

**Reply by marcin-bury** [reply]
Adriano
I will perform more tests - both writeing from node and reading in Delphi app and vice versa.
Since sometimes it shows text properly, sometimes it does not. I will come back with results.

Thanks for the patience
Marcin

**Reply by marcin-bury** [reply]
Probably  I've found the issue, but don't know on which side.
The actual problem is with blob fields. When I enter accented characters on Delphi side, I get the text glitched on node side and vice versa. Although casting blob field to varchar shows proper text in node entered in Delphi. 
Standard varchar fields work ok. 
So I would suspect blob conversion - I used the code from tests:
    for converting to string:

const blobStream = await db.openBlob(transaction, blobColumn);
    const buffer = Buffer.alloc(await blobStream.length);
    await blobStream.read(buffer);
    stringValue = buffer.toString();

for converting from string;

     const blob = await db.createBlob(tr);
     await blob.write(Buffer.from(value));
    await blob.close();      


Is it me doing something wrong or it is something else...

Marcin

**Reply by marcin-bury** [reply]
Hi Adriano

To speed up my learnig curve (I came to node from Delphi world) I decided to convert all the blob fields to long varchar where possible (only a couple of them). But if you don't mind I'd like to ask you for some help resolving this issue. I have one blob field that cannot be converted to varchar and it may contain accented characters.

TIA
Marcin

**Reply by asfernandes** [reply]
Firebird could convert blobs from one charset to another without external libraries (iconv), but this is currently not exposed in the driver. I've created an issue for it https://github.com/asfernandes/node-firebird-drivers/issues/98.

**Comment by iyudincev** [comment]
Use iconv-lite to convert the buffer to string with the use of codepage
other than utf8.

вт, 15 июн. 2021 г. в 10:02, marcin-bury ***@***.***>:

**Reply by marcin-bury** [reply]
Thank you, Igor

This sounds very interesting, I'll give it a try.

Marcin
